sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2169,12 +2169,29 @@ class Analyzer(override val catalogManager: CatalogManager)
                         unbound, arguments, unsupported)
                   }
 
+                  if (bound.inputTypes().length != arguments.length) {
+                    throw QueryCompilationErrors.v2FunctionInvalidInputTypeLengthError(
+                      bound, arguments)
+                  }
+
+                  val castedArguments = arguments.zip(bound.inputTypes()).map { case (arg, ty) =>
+                    if (arg.dataType != ty) {
+                      if (Cast.canCast(arg.dataType, ty)) {
+                        Cast(arg, ty)
+                      } else {
+                        throw QueryCompilationErrors.v2FunctionCastError(bound, arg, ty)
+                      }
+                    } else {
+                      arg
+                    }
+                  }
```

Comment:
As `bound` is come from binding on `UnboundFunction`. From the doc of `bind` method, it works like checking input types.

So that's said, when we bind it, it already checks argument data types, why we need another check and cast here?

Reply:
Yes `bind` checks input types, and potentially allows multiple combinations of them. However, when evaluating the UDF, especially when dealing with magic method, Spark only accept a single set of input parameter types, and so we'll need to insert cast if necessary.

For instance, a UDF can accept both `int` and `decimal` as input types in `bind`, and implements magic method using `decimal` parameters. Spark then should cast `int` arguments to `decimal` when necessary.

Reply:
Since we have two places of type coercion, is there any possibility of bug where we do differently in some cases at both places?

Reply:
Thanks for explaining it. So that is said, `bind` can return a implementation with magic method, which takes decimal input, when Spark binds it with IntegerType input?

Reply:
Then the question is, how the UDF knows Spark can handle the type coercion case?

Using above example, when Spark binds it with IntegerType, the UDF must know Spark can cast int to decimal, so it can return an implementation with magic method taking decimal input?

Reply:
> Since we have two places of type coercion, is there any possibility of bug where we do differently in some cases at both places?

Hmm @dongjoon-hyun what are the two places of type coercion? do you mean the potential inconsistency from `bind` and `inputTypes`?

> Thanks for explaining it. So that is said, bind can return a implementation with magic method, which takes decimal input, when Spark binds it with IntegerType input?

Yes correct. Spark will insert cast `int -> decimal`. 

Also from the Java doc of `inputTypes`: "If the types returned differ from the types passed to `bind(StructType)`, Spark will cast input values to the required data types. This allows implementations to delegate input value casting to Spark."

> Using above example, when Spark binds it with IntegerType, the UDF must know Spark can cast int to decimal, so it can return an implementation with magic method taking decimal input?

Yes the implementor of the UDF should know whether such a cast is valid. Spark also checks this during analysis and throws error if not.

Reply:
It seems to me that a more natural approach is: the v2 function tells Spark all the overloads of its function (e.g. `substring(string)` and `substring(binary)`) , then Spark decides which overload to call and apply necessary type coercions.

Reply:
@cloud-fan I think this approach will not allow us to handle function with varargs, for instance: `max(int, long, float, double, int, long, long, double)`. For this scenario, `bind` can check the input arguments and return a proper implementation of the function (although we still need to do something extra for varargs in Spark).

Reply:
makes sense